chore: allow PrepareModelMigration to be called multiple times

kian99 opened the pull request:
## Description
Migrating a model to JAAS requires a client to call PrepareModelMigration first. This change makes it possible to call PrepareModelMigration multiple times and allows the initial user mapping or backing controller to change. JIMM will ensure that if the model migration has progressed up to a certain point that the PrepareModelMigration will return an error instead to ensure the data is consistent during the migration process.

Fixes [JUJU-8249](https://warthogs.atlassian.net/browse/JUJU-8249)

## Engineering checklist

- [ ] Documentation updated
- [x] Covered by unit tests
- [ ] Covered by integration tests

[JUJU-8249]: https://warthogs.atlassian.net/browse/JUJU-8249?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kian99 committed:
chore: allow PrepareModelMigration to be called multiple times

Migrating a model to JAAS requires a client to call PrepareModelMigration first.
This change makes it possible to call PrepareModelMigration multiple times and allows the initial user mapping or backing controller to change. JIMM will ensure that if the model migration has progressed up to a certain point that the PrepareModelMigration will return an error instead to ensure the data is consistent.

alesstimec commented on internal/db/modelmigration.go:
```diff
@@ -6,15 +6,18 @@ import (
+// if it does not already exist, or updates it if it does.
+// It must be run within a transaction as it will lock the existing row for updates
+// if the model migration already exists.
+func (d *Database) AddOrUpdateIncomingModelMigration(ctx context.Context, modelMigration *dbmodel.IncomingModelMigration) (err error) {
```
this allows for arbitrary update of user mapping.. i suppose this is intended, right?

kian99 replied:
Yip, that's the idea

alesstimec commented on internal/db/modelmigration.go:
```diff
@@ -24,14 +27,60 @@ func (d *Database) AddIncomingModelMigration(ctx context.Context, modelMigration
+		return errors.E(op, "missing uuid", errors.CodeBadRequest)
+	}
+
+	if noWait {
```
how about:
```
lockingClause := clause.Locking{ Strength: "UPDATE" }
if noWait {
  lockingClause.Options = "NOWAIT"
}
db = db.Clauses(lockingClause)
```

kind of reduces the indented bits, bit easier on the eyes

kian99 replied:
Done

kian99 committed:
chore: simplify locking clause

kian99 committed:
chore: prefetch controller

alesstimec commented on internal/db/modelmigration_test.go:
```diff
@@ -87,22 +101,94 @@ func (s *dbSuite) TestGetModelMigration(c *qt.C) {
+				ModelUUID: migration.ModelUUID,
+			}
+			err := d.GetIncomingModelMigrationWithLock(context.Background(), &lookup, noWait)
+			c.Check(err, qt.Equals, nil)
```
are we ok using Check in a goroutine?

kian99 replied:
Yeah check() is okay as it will mark the test as failed but won't stop the running routine if it fails. Assert should be avoided in a go routine because if it fails the go routine will stop.

SimoneDutto commented on internal/jimm/juju/jimm.go:
```diff
@@ -287,7 +287,20 @@ func (j *JujuManager) PrepareModelMigration(
 		}
 
-		if err := j.Database.AddIncomingModelMigration(ctx, &dbmodel.IncomingModelMigration{
+		// Verify the model doesn't exist (implies a migration is in progress
```
the model existing implies the migration is in progress or complete.
Here from the comment it seems the opposite

kian99 replied:
I see the confusion, I'll update the wording.

SimoneDutto commented on internal/jimm/juju/jimm_test.go:
```diff
@@ -804,3 +805,120 @@ func TestPrepareModelMigration_Success(t *testing.T) {
+
+	<-notifyChan // Wait for the migration to be locked
+	_, err = j.PrepareModelMigration(ctx, user, fakeModelUUID, targetController.Name, userMapping)
+	c.Assert(err, qt.IsNotNil)
```
could you test the specific error here?

SimoneDutto reviewed (approved):
just small comments

kian99 committed:
chore: check error matches string

ale8k commented on internal/db/modelmigration.go:
```diff
@@ -24,14 +27,60 @@ func (d *Database) AddIncomingModelMigration(ctx context.Context, modelMigration
+	}
+	// Set noWait to true to ensure that if the row is locked by another transaction,
+	// we will return an error immediately instead of waiting.
+	err = d.GetIncomingModelMigrationWithLock(ctx, &lookup, true)
```
Can we do the adding and updating explicitly please? So in calls to add or update, they'll attempt to get it first? We often do that.

kian99 replied:
I'm confused on this one, we are getting it first?

ale8k commented on internal/db/modelmigration.go:
```diff
@@ -6,15 +6,18 @@ import (
+// It must be run within a transaction as it will lock the existing row for updates
+// if the model migration already exists.
+func (d *Database) AddOrUpdateIncomingModelMigration(ctx context.Context, modelMigration *dbmodel.IncomingModelMigration) (err error) {
 	const op = errors.Op("db.AddIncomingModelMigration")
```
Op

ale8k commented on internal/db/modelmigration.go:
```diff
@@ -6,15 +6,18 @@ import (
-func (d *Database) AddIncomingModelMigration(ctx context.Context, modelMigration *dbmodel.IncomingModelMigration) (err error) {
+// AddOrUpdateIncomingModelMigration stores information about an incoming model migration
+// if it does not already exist, or updates it if it does.
+// It must be run within a transaction as it will lock the existing row for updates
```
If it must always be a transaction, I think we can just start a transaction in this method. You can nest transactions and they just act as savepoints.

ale8k commented on internal/db/modelmigration.go:
```diff
@@ -24,14 +27,60 @@ func (d *Database) AddIncomingModelMigration(ctx context.Context, modelMigration
 	db := d.DB.WithContext(ctx)
 
-	if err := db.Create(modelMigration).Error; err != nil {
+	if err := db.Save(modelMigration).Error; err != nil {
```
IIRC using db.Model(x).Save(...) can lead to no updated_at field beign set, I'm guessing this is fine?

ale8k commented on internal/db/modelmigration_test.go:
```diff
@@ -75,7 +89,7 @@ func (s *dbSuite) TestGetModelMigration(c *qt.C) {
 
 	lookup := dbmodel.IncomingModelMigration{ModelUUID: migration.ModelUUID}
-	err = s.Database.GetIncomingModelMigration(context.Background(), &lookup)
+	err := s.Database.GetIncomingModelMigration(context.Background(), &lookup)
```
Please check it looks as intended here, we're just checking err but perhaps it actually gotten nothing

ale8k commented on internal/db/modelmigration_test.go:
```diff
@@ -44,28 +50,36 @@ func (s *dbSuite) TestAddModelMigration(c *qt.C) {
-	migration.ID = 0 // Reset ID to ensure it is not reused
-	err = s.Database.AddIncomingModelMigration(context.Background(), &migration)
-	c.Assert(err, qt.Not(qt.IsNil))
+	c.Assert(dbMigration.UpdatedAt, qt.Not(qt.Equals), time.Time{})
```
Ahhh you checked the updated_at, scratch my other comment. Ty for this!

ale8k commented on internal/jimm/juju/migrationtarget.go:
```diff
@@ -420,26 +420,41 @@ func (j *JujuManager) Import(ctx context.Context, user *openfga.User, serialized
-	if err != nil {
-		return errors.E(op, fmt.Errorf("failed to add incoming model migration: %w", err))
-	}
+		// Set noWait to false to allow the transaction to wait for the lock.
```
Why false? Would we ever import the same migration multiple times and want it to queue?

kian99 replied:
Not quite, it's simply that if another routine is holding the lock, the import method should wait to grab it rather than error out.

ale8k commented on internal/jimm/juju/jimm.go:
```diff
@@ -287,7 +287,20 @@ func (j *JujuManager) PrepareModelMigration(
+		model := &dbmodel.Model{
+			UUID: sql.NullString{String: modelUUID, Valid: true},
+		}
+		err := j.Database.GetModel(ctx, model)
```
Could make a little func wrapping this called:

"CanPrepareModelMigration" or something, then you can put this doc there. Cause really it's just checking that it's safe to Prepare a model migration and can be godoc'd on a func.

kian99 replied:
I think that's the same thing but hides the logic inside a function. Without needing to reuse that function I'll keep this here as is

ale8k reviewed (approved):
LGTM bar some minor things. But can we not have AddOrUpdate as a method. I'd prefer just Add and Update as separate methods (makes the code clearer). Having AddOrUpdate can lead to some funky stuff.

Can we also just put the TX within the Add/Update and not have godoc saying it must be run in a transaction. If it must be run in a transaction, place it in a transaction aha.

kian99 committed:
chore: run db method with nested transaction

ale8k commented on internal/db/modelmigration.go:
```diff
@@ -27,25 +25,29 @@ func (d *Database) AddOrUpdateIncomingModelMigration(ctx context.Context, modelM
-	} else if err != nil && errors.ErrorCode(err) != errors.CodeNotFound {
-		return errors.E(op, err)
-	}
+	err = d.Transaction(func(d *Database) error {
```
Ty!

kian99 committed:
chore: use the tx in a transaction

It's easy to accidentally use j.Database when inside a transaction closure, but this means your DB query may use a different connection to the database and not the one where you have started a transaction.
This change fixes that in a few places.
